lib/custom_image_crop_widget.dart

```diff
@@ -0,0 +1,255 @@
+  }
+
+  void onMoveUpdate(MoveEvent event, Path path) {
+    addTransition(CropImageData(x: event.delta.dx, y: event.delta.dy));
+  }
+
+  Path getPath(double cropWidth, double width, double height) {
+    if (widget.path != null) {
+      return widget.path(cropWidth);
+    }
+    switch (widget.shape) {
+      case CustomCropShape.Circle:
+        return Path()
+          ..addOval(
+            Rect.fromCircle(
+              center: Offset(width / 2, height / 2),
+              radius: cropWidth / 2,
+            ),
+          );
+      default:
+        return Path()
+          ..addRect(
+            Rect.fromCenter(
+              center: Offset(width / 2, height / 2),
+              width: cropWidth,
+              height: cropWidth,
+            ),
+          );
+    }
+  }
+
+  @override
+  Future<MemoryImage> onCropImage() async {
+    if (imageAsUIImage == null) {
+      return null;
+    }
+    final cropWidth = min(width, height) * widget.cropPercentage;
+    final pictureRecorder = ui.PictureRecorder();
+    final canvas = Canvas(pictureRecorder);
+    final defaultScale = min(imageAsUIImage.width / cropWidth, imageAsUIImage.height / cropWidth);
+    final scale = data.scale * defaultScale / 2;
+    final clipPath = Path.from(getPath(cropWidth, cropWidth, cropWidth));
+    final matrix4Image = Matrix4.diagonal3(vector_math.Vector3(1, 1, 0))
+      ..translate(data.x + cropWidth / 2, data.y + cropWidth / 2)
+      ..scale(scale)
+      ..rotateZ(data.angle);
+    final bgPaint = Paint()
+      ..color = widget.backgroundColor
+      ..style = PaintingStyle.fill;
+    canvas.drawRect(Rect.fromLTWH(0, 0, cropWidth, cropWidth), bgPaint);
+    canvas.save();
+    canvas.clipPath(clipPath);
+    final paint = Paint()..isAntiAlias = false;
+    canvas.transform(matrix4Image.storage);
+    canvas.drawImage(imageAsUIImage, Offset(-imageAsUIImage.width / 2, -imageAsUIImage.height / 2), paint);
+    canvas.restore();
+    ui.Picture picture = pictureRecorder.endRecording();
+    ui.Image image = await picture.toImage(cropWidth.floor(), cropWidth.floor());
+
+    // Optionally remove magenta from image by evaluating every pixel
+    // See https://github.com/brendan-duncan/image/blob/master/lib/src/transform/copy_crop.dart
+
+    ByteData bytes = await image.toByteData(format: ui.ImageByteFormat.png);
+    return MemoryImage(bytes.buffer.asUint8List());
+  }
+
+  @override
+  void addTransition(CropImageData transition) {
+    setState(() {
+      if (dataAllowed(data + transition)) {
+        data += transition;
+      }
+    });
+  }
+
+  bool dataAllowed(CropImageData data) {
+    // For now, this will do. The idea is that we create
+    // a path from the data and check if when we combine
+    // that with the crop path that the resulting path
+    // overlap the hole (crop). So we check if all pixels
+    // from the crop contain pixels from the original image
+    data.scale = data.scale.clamp(0.1, 10.0);
```

Comment:
This does nothing given that data+transition will create a new object that is destroyed afterwards

Reply:
Also the naming of this function does not convey that it will change the data